File: myFinance_improved_V2/lib/features/inventory_management/data/models/product_model.dart
// Data Model: Product Model (Compatibility Layer)
// This file maintains backward compatibility while delegating to DTO and Mapper
// New code should use ProductDto and ProductMapper directly

import '../../domain/entities/product.dart';
import '../dtos/product_dto.dart';
import '../mappers/product_mapper.dart';

/// @deprecated Use ProductDto and ProductMapper instead
/// This class is maintained for backward compatibility only
class ProductModel extends ProductDto {
  const ProductModel({
    required super.id,
    required super.name,
    super.sku,
    super.barcode,
    required super.price,
    super.cost,
    required super.stock,
    super.minStock,
    super.maxStock,
    super.unit,
    super.categoryId,
    super.categoryName,
    super.brandId,
    super.brandName,
    super.productType,
    super.description,
    super.imageUrls,
    super.isActive = true,
    super.stockStatus,
    super.quantityAvailable,
    super.quantityReserved,
    super.createdAt,
    super.updatedAt,
    // v6 variant fields
    super.variantId,
    super.variantName,
    super.variantSku,
    super.variantBarcode,
    super.displayName,
    super.displaySku,
    super.displayBarcode,
    super.hasVariants = false,
  });

  // From JSON - delegates to ProductDto
  factory ProductModel.fromJson(Map<String, dynamic> json) {
    final dto = ProductDto.fromJson(json);
    return ProductModel(
      id: dto.id,
      name: dto.name,
      sku: dto.sku,
      barcode: dto.barcode,
      price: dto.price,
      cost: dto.cost,
      stock: dto.stock,
      minStock: dto.minStock,
      maxStock: dto.maxStock,
      unit: dto.unit,
      categoryId: dto.categoryId,
      categoryName: dto.categoryName,
      brandId: dto.brandId,
      brandName: dto.brandName,
      productType: dto.productType,
      description: dto.description,
      imageUrls: dto.imageUrls,
      isActive: dto.isActive,
      stockStatus: dto.stockStatus,
      quantityAvailable: dto.quantityAvailable,
      quantityReserved: dto.quantityReserved,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt,
      // v6 variant fields
      variantId: dto.variantId,
      variantName: dto.variantName,
      variantSku: dto.variantSku,
      variantBarcode: dto.variantBarcode,
      displayName: dto.displayName,
      displaySku: dto.displaySku,
      displayBarcode: dto.displayBarcode,
      hasVariants: dto.hasVariants,
    );
  }

  // Mapper: Model → Entity - delegates to ProductMapper
  Product toEntity() {
    return ProductMapper.toEntity(this);
  }

  // Mapper: Entity → Model - delegates to ProductMapper
  factory ProductModel.fromEntity(Product entity) {
    final dto = ProductMapper.fromEntity(entity);
    return ProductModel(
      id: dto.id,
      name: dto.name,
      sku: dto.sku,
      barcode: dto.barcode,
      price: dto.price,
      cost: dto.cost,
      stock: dto.stock,
      minStock: dto.minStock,
      maxStock: dto.maxStock,
      unit: dto.unit,
      categoryId: dto.categoryId,
      categoryName: dto.categoryName,
      brandId: dto.brandId,
      brandName: dto.brandName,
      productType: dto.productType,
      description: dto.description,
      imageUrls: dto.imageUrls,
      isActive: dto.isActive,
      stockStatus: dto.stockStatus,
      quantityAvailable: dto.quantityAvailable,
      quantityReserved: dto.quantityReserved,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt,
      // v6 variant fields
      variantId: dto.variantId,
      variantName: dto.variantName,
      variantSku: dto.variantSku,
      variantBarcode: dto.variantBarcode,
      displayName: dto.displayName,
      displaySku: dto.displaySku,
      displayBarcode: dto.displayBarcode,
      hasVariants: dto.hasVariants,
    );
  }
}

File: website/src/features/sale-product/data/models/ProductModel.ts
/**
 * ProductModel
 * DTO and Mapper for Product entity
 */

import { Product } from '../../domain/entities/Product';

export interface ProductDTO {
  product_id: string;
  sku: string;
  product_name: string;
  brand_name: string;
  category_name: string;
  unit: string;
  image_urls: string[];
  price: {
    selling: number;
  };
  stock: {
    quantity_available: number;
  };
}

/**
 * ProductForSale interface
 * Used in presentation layer for product listing
 */
export interface ProductForSale {
  product_id: string;
  product_name: string;
  sku: string;
  barcode: string;
  brand_name: string;
  category_name: string;
  unit: string;
  image_urls: string[];
  selling_price: number;
  quantity_available: number;
}

export class ProductModel {
  /**
   * Convert DTO to Domain Entity
   */
  static toDomain(dto: ProductDTO): Product {
    return Product.create({
      id: dto.product_id,
      sku: dto.sku,
      name: dto.product_name,
      brandName: dto.brand_name,
      categoryName: dto.category_name,
      unit: dto.unit,
      imageUrls: dto.image_urls || [],
      sellingPrice: dto.price?.selling || 0,
      quantityAvailable: dto.stock?.quantity_available || 0,
    });
  }

  /**
   * Convert Domain Entity to DTO
   */
  static fromDomain(product: Product): ProductDTO {
    return {
      product_id: product.id,
      sku: product.sku,
      product_name: product.name,
      brand_name: product.brandName,
      category_name: product.categoryName,
      unit: product.unit,
      image_urls: product.imageUrls,
      price: {
        selling: product.sellingPrice,
      },
      stock: {
        quantity_available: product.quantityAvailable,
      },
    };
  }

  /**
   * Convert array of DTOs to Domain Entities
   */
  static toDomainList(dtos: ProductDTO[]): Product[] {
    return dtos.map((dto) => ProductModel.toDomain(dto));
  }

  /**
   * Convert ProductForSale (presentation layer) to Domain Entity
   * Used when adding products to cart from the product list
   */
  static fromSaleProduct(saleProduct: ProductForSale): Product {
    return Product.create({
      id: saleProduct.product_id,
      sku: saleProduct.sku,
      name: saleProduct.product_name,
      brandName: saleProduct.brand_name,
      categoryName: saleProduct.category_name,
      unit: saleProduct.unit,
      imageUrls: saleProduct.image_urls || [],
      sellingPrice: saleProduct.selling_price,
      quantityAvailable: saleProduct.quantity_available,
    });
  }
}
